### DISCUSSION TITLE
CI/CD recommendations/chapter

### DISCUSSION THREAD

**Discussion by tgmendes** [BretFisher/ama#124] [Q&A]
Hello Brett,

My question is related on the ideal setup of a CI/CD for docker build and deployment. Currently we are focusing only on open source solutions for a CI (we use Jenkins), and we have our own personal registry for docker images.

So my question is, **what is the recommended pipeline for image building, pushing and updating**?

Let me ellaborate further on what we have and what are my doubts. Due to infrastructure related issues, we have Jenkins on a separate machine from the production/staging servers. Here's the flow:

1. When we push code to our Dev or Master branch, it triggers a build in Jenkins;
2. This build will do integration tests, build the relevant images (using the docker plugin) and push them to our repo;
3. The production compose files are copied to the remote servers;
4. Jenkins issues an SSH command to the remote server to update the application with the new builds (right now we use simply compose, but we'll switch to swarm in the near future).

Question is, is point **3** and **4** the right way to do things? I know you can run docker commands on a remote machine without the need of using SSH commands, but I still haven't explored this possibility.

This leads my to a **suggestion** for the Docker mastery course: why not have a more detailed explanation on a typical setup for CI/CD with Docker? You do talk about it, and mention some solutions, but I think it could be useful to rurther ellaborate.

Thanks a lot,
Tiago

**Comment by BretFisher** [comment]
Hi Tiago. Some quick answers:

- 3 and 4 are fine. Once you get to Swarm Stacks you should be able to stream the compose file over in a single SSH command to save steps. Having secure Docker API on TCP for remote docker cli is really just a preference IMO. If SSH works for you, then the remote cli isn't going to give you new abilities, it's just different (TLS) keys you'd need to manage and requires some extra setup to secure the docker engine API.
- Once you're on Swarm, look at cloud.docker.com "bring your own swarm", which sets up the remote API TLS for you. I've never used that in Jenkins for reasons above, but I do use it for remote admin from my local machine.
- The one benefit of using remote API is you don't have to transfer yml files for Swarm Stacks to the server, but with the new stream ability of stack deploy command it's not much different for CI/CD deployments.
- I've had lots of requests for CI/CD but it would take its own multi-hour course to go into detail on multiple tools. The CI/CD scene is quite diverse, so to do it justice, I plan on making a smaller course just for those topics... just not sure when exactly :)

**Comment by tgmendes** [comment]
Thanks a lot for the insights Bret. If you ever visit Portugal I'll buy you a beer :)

I would look forward to see such course, but I understand it's not easy (mainly because there are a lot of CI/CD systems with their own nuances to cover). 

Regards,
Tiago

**Comment by BretFisher** [comment]
Some more info from a recent email:

If you're trying to accomplish "full auto CI/CD" to a Swarm, a few hints:

You can pipe a compose file into the stack update command... so you won't need to copy files to server. You could put all that into a ssh command I think, that CI could run once image is uploaded to the registry:

`cat docker-compose.yml | ssh user@server "cat > docker stack deploy --compose-file - mystack"`

Notice the `-` in the stack deploy command which means pull yaml from stdin.  

One potential issue is that I recommend not using "latest" tag, and to version all images, but then that means your stack file needs to update each time.  You can do taht by setting envvars in your CI and have the compose file pull in that envvar, so steps in CI, triggered by a code commit, could be:

1. build image with tag, maybe commit-id, maybe date/time, etc. 
2. set that special tag to envvar
2. if success, test in image
3. if success, push image to a registry
4. run example cat+ssh+stack deploy command above with variables.

If you didn't want CI doing all that, you could have your Swarm running a Service with a custom container monitoring for a Docker Hub webhook and then it would kick off a stack deploy.

Here's an [example using incoming webhooks](https://blog.viktoradam.net/2018/01/13/home-lab-part3-swarm-cluster/#updatingstacks) and the [repo he created](https://github.com/rycus86/webhook-proxy).

There are other tools out there like that, but most of them do a manual check of the latest image (pull method) rather than wait for Docker Hub webhook (push method). Each has its use based on your requirements.
